src/cyrsasl_external.erl

```diff
@@ -0,0 +1,61 @@
+-module(cyrsasl_external).
+-author('denys.gonchar@erlang-solutions.com').
+
+-xep([{xep, 178}, {version, "1.1"}, {comment, "partially implemented."}]).
+
+-export([start/1, stop/0, mech_new/2, mech_step/2]).
+
+-behaviour(cyrsasl).
+
+-record(state, {creds}).
+
+start(_Opts) ->
+    cyrsasl:register_mechanism(<<"EXTERNAL">>, ?MODULE, cert),
+    ok.
+
+stop() ->
+    ok.
+
+-spec mech_new(Host :: ejabberd:server(),
+               Creds :: mongoose_credentials:t()) -> {ok, tuple()}.
+mech_new(_Host, Creds) ->
+  case mongoose_credentials:get(Creds,client_cert,undefined) of
+    undefined -> {ok, #state{creds = invalid_cert}};
+    Cert ->
+      DerCert = public_key:pkix_encode('Certificate', Cert, plain),
+      PemCert = public_key:pem_encode([{'Certificate', DerCert, not_encrypted}]),
+      ExtraFields = get_common_name(Cert) ++ get_xmpp_addresses(Cert),
+      {ok, #state{creds = mongoose_credentials:extend(Creds, [ {pem_cert, PemCert},
```

Comment:
`[ {pem_cert, .....` can go into a separate variable. ExtraFields2 or something.

Reply:
there is dedicated variable  ExtraFields, because it can be [ ], or list of definitions. and that is not the case for DerCert and PemCert.

Reply:
OK

```
ExtraFields = get_common_name(Cert) ++ get_xmpp_addresses(Cert),
AllFields = [{pem_cert, PemCert}, {der_cert, DerCert} | ExtraFields],
mongoose_credentials:extend(Creds, AllFields)
```

Reply:
I need to know that the second argument of extend is fields just by looking at the code.

Reply:
IMO it is fine as it is now and looks clear. :)